### ISSUE TITLE
Enhancement: [no-unnecessary-condition] detect unnecessary type guard callbacks passed directly

### ISSUE THREAD

**Issue by g-drouard** [typescript-eslint/typescript-eslint#12343]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### Description

When a type guard function is passed directly as a callback (e.g., `arr.filter(isNotNil)`), the rule currently does not check whether the predicate is always satisfied or can never be satisfied.

It only detects the inline form `arr.filter(x => isNotNil(x))`.

### Fail

```TypeScript
// Always truthy — type guard is unnecessary
declare function isNotNil<T>(value: T | null | undefined): value is T;
declare const arr: string[];
arr.filter(isNotNil); // string is already non-nullable

// Always truthy — via RxJS-style operator
declare const obs: Observable<string>;
obs.pipe(filter(isNotNil));

// Always falsy — no overlap between element type and predicate type
declare function isString(value: unknown): value is string;
declare const numbers: number[];
numbers.filter(isString); // number has no overlap with string

// Always falsy — isNil on non-nullable array
declare function isNil(value: unknown): value is null | undefined;
declare const names: string[];
names.filter(isNil);

// Always truthy — type guard at non-first argument position
declare function filterWith<T, S extends T>(source: T[], predicate: (value: T) => value is S): S[];
filterWith(arr, isNotNil);
```

### Pass

```TypeScript
// Necessary — array has nullable elements
declare function isNotNil<T>(value: T | null | undefined): value is T;
declare const arr: (string | null)[];
arr.filter(isNotNil);

// Necessary — predicate narrows a subtype
declare function isString(value: unknown): value is string;
declare const mixed: (string | number)[];
mixed.filter(isString);

// Necessary — RxJS with nullable observable
declare const obs: Observable<string | null>;
obs.pipe(filter(isNotNil));

// Not a predicate-consuming API — should not report
declare const numbers: number[];
numbers.map(isString);
numbers.forEach(isString);

// any[] / unknown[] — can't determine, should not report
declare const anyArr: any[];
anyArr.filter(isNotNil);

// Non-type-guard callback — should not report
declare function registerCallback(cb: (value: string) => boolean): void;
registerCallback(isNotNil);
```

**Comment by g-drouard** [comment]
I opened a [draft PR](https://github.com/typescript-eslint/typescript-eslint/pull/12344), I'm waiting for the "accepting PRs" flag.

**Comment by kirkwaiblinger** [comment]
Interesting case! 

I'd like to separate this into two pieces

1. Detect unnecessary type predicates in array callbacks of the form `arr.filter(typeGuard)`, using the same logic that currently flags an unnecessary type guard of the form `arr.filter(x => typeGuard(x))`. I'm +1 to accepting this proposal (especially since we already do the same thing with things like `arr.filter(alwaysReturnsTrue)`)

2. Adding additional detection of unnecessary callbacks where the result is never possible, such as `numbers.filter(isString)`. This portion of the proposal is a subset/duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/10997, so we should not consider it as part of this issue.

---

Also, please read our [AI contribution policy](https://typescript-eslint.io/contributing/ai-policy/). This issue is quite verbose and contains redundant/irrelevant examples, but it's ok to leave it as is for now. The draft PR, however, is overtly in violation of the policy and won't be reviewed as-is, so, please rework that if you intend to undraft it. Thank you!

**Comment by kirkwaiblinger** [comment]
Also - I don't think that this case is something we want to tackle:
```ts
// Always truthy — type guard at non-first argument position
declare function filterWith<T, S extends T>(source: T[], predicate: (value: T) => value is S): S[];
filterWith(arr, isNotNil)
```

Just things of the form `arr.filter(callback)` and of course the rest of the array callback methods for which the rule already detects functions which always return a truthy value:

https://github.com/typescript-eslint/typescript-eslint/blob/63ab00225eea8c9e4c8fadc22d5c3081daf13e18/packages/eslint-plugin/src/util/isArrayMethodCallWithPredicate.ts#L13-L21

**Comment by Arlikhozhaev** [comment]
I’m going to work on Kirk’s part 1: array methods like filter/find, when the type guard is already unnecessary, same as arr.filter(x => isNotNil(x)). I won’t add always-false predicates, RxJS, or filterWith. Seems like #12344 is now closed, so I’ll open a new PR.
